pypeit/core/tracewave.py

```diff
@@ -244,12 +241,29 @@ def trace_tilts_work(arcimg, lines_spec, lines_spat, thismask, slit_cen, inmask=
         sub_img = arcimg_trans[min_spat:max_spat, :]
         sub_inmask = inmask_trans[min_spat:max_spat,:]
         sub_thismask = thismask_trans[min_spat:max_spat,:]
-        if do_crude: # First time tracing, do a trace crude
-            tilts_guess_now, err_now = trace_slits.trace_crude_init(
-                sub_img, np.array([lines_spec[iline]]), (sub_img.shape[0] - 1) // 2, invvar=sub_inmask, radius=fwhm,
-                nave=tcrude_nave, maxshift0=tcrude_maxshift0, maxshift=tcrude_maxshift, maxerr=tcrude_maxerr)
-            tilts_guess_now=tilts_guess_now.flatten()
-        else:  # A guess was provided, use that as the crutch, but determine if it is a full trace or a sub-trace
+        if do_crude:
+            # First time tracing, do a trace crude
+
+            # TODO: Worried that using a starting row that is blindly
+            # set as half of sub_img is unstable to very curved slit
+            # traces. Should instead start at the center of the slit as
+            # defined by the slit trace?
+
+            # NOTE: follow_centroid behaves differently from the old
+            # trace_crude_init within 2-4 pixels at the trace edge
+
+            # Smooth the image
+            tilts_guess_now, err_now, _ \
```

Comment:
follow_centroid needs to be modified to take nave as an input parameter as was done in the original code trace_crude_init that was ported. You can do that by simply calling your boxcar smoothing function that you added in utils.  Note that in the original code nave=5 was the default. If the code does not do some boxcar smoothing, it is not going to work, which is why I think this needs to be in the code itself, and not something you perform on the arguments. 

The other point is that your utils.boxcar_smooth_rows as it is implemented here is not using the mask, whereas the original implementation in trace_crude_init is using the mask (passed in as the invvar). That is probably why you are getting different (and possibly spurious) results near the edges. 

Please ensure that your new function utils.boxcar_smooth_rows is: 

1) called in follow_centroid as the default behavior as was the case in trace_crude_init
2) is taking the mask/invvar when performing the convolution, as was the original behavior. 
3) produces the same results as trace_crude_init everywhere.

Reply:
1. I disagree with putting the call to `utils.boxcar_smooth_rows` inside `trace.follow_centroid`.  I see it as maintaining the modularity of the code, but I'm also not following why you think `trace.follow_centroid` would not work without the smoothing; i.e., I don't see the smoothing as a core functionality of `trace.follow_centroid`.  As it is, it's having to do this smoothing once per line; naively it seems like we should need to do this far fewer times (possibly only once), which means we want to run the smoothing beforehand and then pass a sub image of the result to `trace.follow_centroid`.

2. You're right that I did mistakenly miss out the mask.  Thanks for finding this error!  I've added the mask back in.  

3. For slit-edge tracing, I did a lot of testing of the initial tracing in the new code compared to the `trace_crude_init` code; however, it was never a check that the results were identical.  I'm running some tests now with the tilt tracing, and the results are still slightly different and I'm still tracking down why.  However, if you have time, I'd really appreciate it if you would run the DevSuite on a couple of data sets where the tilt and object tracing are tricky or particularly important and check the results of the current code.  Personally (perhaps controversially), I'm less concerned that the behavior of the new code can identically reproduce the old code, and more concerned that the new code is producing valuable results.

Reply:
trace_crude does not work well if the images are not smoothed is my experience, which is why I think a default of smoothing of 5 pixels is a reasonable choice. The smoothing is only performed once per image. By separating the codes in the way you have done them, the user will likely forget about the smoothing, whereas the smoothing should be the default. 

The smoothing only takes place one per call, so whether it occurs outside the routine or inside the routine is irrelevant to performance. That said, the tilts code is indeed doing this smoothing many times on sub-images. That is true in the previous code and your changes to it. More efficient would be to do this smoothing once, but I would need to do some tests to make sure that this change does not break things. That said, this could be easily accommodated by simply setting nave = None. 

@profxj ported trace_crude without adding this smoothing, and for that reason it was working poorly. I then realized this and added that smoothing, which is why I think the smoothing needs to be the default. The code remains modular, but the point is the user is not able to run the code without the wrong settings, which is what your change will encourage.

Reply:
I don't see why your follow_trace plus boxcar smoothing would not reproduce exactly the same results as trace_crude_init. The point is that you stated that trace_fweight and moment1d produce identical results. trace_crude_init calls trace_fweight, and follow_trace must call moment1d, so I fail to see why the result would not be identical. If they are not, there is likely a bug somewhere related to masking. I believe that I found at least one such bug, so I feel that if the result are not identical that is dangerous, because you are probably not testing the edge cases where the masking is important.

Reply:
Hmm... I'm now wondering if it's the error limit, `tcrude_maxerr`.  I fixed the error calculations, so it may be that the value of `tcrude_maxerr=1.0` needs to be recalibrated.  Were the results of the tracing sensitive to this parameter?  Just to say again, the results are different but it's difficult for me to tell if they're dramatically different.  E.g., the RMS values in the tilt tracing QA plots are comparable.